src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -317,7 +317,7 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
                 // for sets and read-only set interfaces, we clone what's there into a new collection, if we can
                 if (TypeIsASetInterface(type) && !bindingPoint.IsReadOnly)
                 {
```

Comment:
```suggestion
                if (TypeIsASetInterface(type))
                {
                    if (bindingPoint.IsReadOnly && bindingPoint.Value is null)
                    {
                        return;
                    }
```

This is very much an edge-case, but with the current logic, given an `public ISet<string> MySet { get; } = new MyCustomSet();`, the custom `ICollection<string>.Add(T item)` will be called while with `public ISet<string> MySet { get; set; } = new MyCustomSet();`, the custom `ISet<string>.Add(T item)` which is a different method which returns a bool.

Another option to keep maximum compatability with pre-.NET 7 code would be to always use `ICollection<T>.Add` when populating an `ISet<T>`, but either way we should be consistent with which `Add` method we use populate non-null `ISets<T>`.